rfcs/20190506-filesystem-plugin-modular-tensorflow.md

```diff
@@ -0,0 +1,1082 @@
+
+Since this has to be pure C, Status objects are replaced by pointers to a
+`TF_Status` structure. There is code to convert from one to the other. Moreover,
+these APIs assume that the `TF_Status` object is created and managed from the
+calling code, instead of returning the status from every function call. We will
+follow a similar design in this proposal.
+
+## Proposed implementation
+
+Since the existing implementation is layered, as described above, we can focus
+this design on the low level API part. Everything else can be incorporated into
+the core of TensorFlow in the modular world and it should support any filesystem
+in a transparent way (as long as the plugin is loaded).
+
+### Filesystem interface
+
+In this section, we present the interface that all filesystem plugins must
+implement before they can be loaded and used.
+
+Across core-plugin boundary, we can represent open files and a filesystem as
+opaque pointers (`void*`) but this results in a loss of type safety. It is easy
+to imagine passing a pointer to a read only file to a method expecting to write
+to such a file. To prevent this from happening, we propose the following 4 data
+structures which just incorporate the `void*` pointer:
+
+```cpp
+typedef struct TF_RandomAccessFile {
+  void* plugin_file;
+} TF_RandomAccessFile;
+
+typedef struct TF_WritableFile {
+  void* plugin_file;
+} TF_WritableFile;
+
+typedef struct TF_ReadOnlyMemoryRegion {
+  void* plugin_memory_region;
+} TF_ReadOnlyMemoryRegion;
+
+typedef struct TF_Filesystem {
+  void* plugin_filesystem;
+} TF_Filesystem;
+```
+
+The glue code between the core TensorFlow and the plugin will
+allocate/deallocate these structures but they can also be allocated/deallocated
+by plugins during testing, for example:
+
+```cpp
+class PluginFilesystemIOTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    filesystem_ = new TF_Filesystem;
+    Init(filesystem_);
+    status = TF_NewStatus();
+  }
+
+  void TearDown() override {
+    TF_DeleteStatus(status);
+    Cleanup(filesystem_);
+    delete filesystem_;
+  }
+
+  TF_Filesystem* filesystem_;
+  TF_Status* status;
+};
+
+TEST_F(PluginFilesystemIOTest, TestIO) {
+  // ...
+}
+```
+
+These structures are useless without operations to operate on them. We could
+incorporate function pointer in the structures themselves but that will result
+in a lot of duplication and object size changes when one functionality gets
+added/removed. This, in turn results in compatibility being harder to maintain,
+which is something we don’t want.
+
+Instead, we provide 3 different data structures to act as function tables, to
+hold function pointers for files and the filesystem. Currently, these structures
+are filled with the functionality already present in the existing filesystem
+API:
+
+```cpp
+// Operations on a TF_RandomAccessFile
+typedef struct TF_RandomAccessFileOps {
+  size_t (*const Read)(const TF_RandomAccessFile*, uint64, size_t, char*, TF_Status*);
```

Comment:
@mihaimaruseac One very useful method for random access file could be `Size` or `GetSize` which returns the size of the file. At the moment, File Size is not visible from RandomAccessFile which is really inconvenient in many situations.

The reason being that many file format actually has the meta data at the end of the file (not at the beginning of the file). One good (and important) example is the zip file. Zip file's meta data is stored at the end of the file (central directory). So any parser to open a zip file will have to find the file size first before processing.

With a `GetSize` (or `Size`) API, RandomAccessFile could be greatly optimized.

Note technically it is possible to find the file size with `Read` to probe through until the EOF is reached, but this is really not efficient.
 
Of course, some file systems may not be able to provide a `Size` before hand (e.g., in stream fashion). I think by default `GetSize` could simply return `NotImpemented`. So any file size based optimization will only be "best effort".

For optimization, if a `GetSize`/`Size` API is exposed, then reading files could try `GetSize` first. If `NotImplemented` is returned, then a normal approach of probe through until the end is used. Otherwise the size of the file could be directly used.

Note: ZIP is the format mentioned but a lot of the files are actually zip files with `[Content_Types].xml` or `META-INFO` file inside.

For example, Windows office files (pptx, docx, etc) are all zip files. Java jar files are zip files. And python whl files are zip files as well.

You could just use unzip to decompress pptx, jar, whl actually and see each individual content files.

Reply:
I agree with this request. I only included the API calls that are currently included in the existing classes but I guess we can expand these. I'm going to add a new section on proposed additions to the op tables.

Reply:
Actually one of the fields of `Read` is the offset from where you want to read and the other is the number of bytes to read. So, if reading a zip file, you can go directly to the end of the file and read the size section from there.

Sure, we can add it as convenience function but we only identify the filesystem plugin based on the scheme part of the URI, not the extension (`hdfs://path/to/file.zip` and `/path/to/file.zip` will use different implementations but `/path/to/file.zip` and /path/to/another/file.txt` will use the same ops)

Reply:
@mihaimaruseac while it is possible to read the content with offset, the `Read` API does not give you the option to specify `offset = -10` to read from 10 bytes before the end of the file EOF. In oder to achieve that, we will need to find out the length of the file (e.g., say 100), and subtract by 10 (100 - 10 = 90) to read from offset `90` (vs. `-10`).

The filesystem plugin will be consumed by a client calling it (a function inside tensorflow, or another plugin/customer ops outside of tensorflow repo).

For example, inside tensorflow one function might want to parse a file (not necessary zip file) with file name `hdfs://path/to/special/file`. Outside of tensorflow another plugin want to access Tensorflow's file system schema URI `gcs://path/to/another/file`.

Eventually the  function (inside or outside tensorflow) will need to access `Read` (or proposed `GetSize/GetFileSize`) through the plugin.

If `GetSize/GetFileSize` is not exposed, then size is unknown to this function (function inside tensorflow, or another plugin using tensorflow). This function will have to probe through to find the EOF.

Reply:
I wasn't clear enough, sorry. Let me try with an example.

Suppose we live in a world where there are only 3 types of files: `*.start` which have the size metadata at the beginning, `*.end` which have it at the end and `*.none` which have no size metadata.

Suppose we also have only two filesystems: A (`a:///path/to/file`) and B (`b:///path/to/file`). Let's assume that A allows rewinds (so you can read the file and then come back to its start) while B is only streaming (once you read some part of the file, you can't go back, unless you open the file again).

Suppose we have a piece of user code which takes as input a filepath and returns the size of the file, let's call this `ExternalFileSize(const string& path)`.

There are 4 posibilities:

1. We keep the API as it is. In this case, `ExternalFileSize(path)` will call `Env::Default()->NewRandomAccessFile(...)` to get a new `RandomAccessFile` object and will have to call `Read` on it until the end increasing a size variable and then returning it. This, I agree with you, is not that good.

    ```cpp
    // This code is outside TensorFlow
    size_t ExternalFileSize(const string& path) {
      RandomAccessFile file;
      Env::Default()->NewRandomAccessFile(path, &file);
      return ReadFromStartToEndAndGetSize(file);
   }

   // Extracting this as separate function as I refer to it below
   // This is still outside TensorFlow
   ReadFromStartToEndAndGetSize(RandomAccessFile file) {
      size_t size = 0;
      // ignore casts for size_t to uint64 and from Status to bool
      while (file->Read(size, amount_to_read, output, result)) {
        // assume no partial reads for now as it doesn't matter here
        size += amount_to_read;
      }
      return size;
    }
    ```

2. We add a `GetSize` method in the plugin API and in `RandomAccessFile` class. To implement `ExternalFileSize(path)` we now call `NewRandomAccessFile` and `GetSize` on the returned object. However, `NewRandomAccessFile` dispatches on the scheme, so both plugins need to handle `*.start`, `*.none` and `*.end` files inside the `GetSize` equivalent, resulting in code duplication as well as behavior that is not common across plugins. This is what I was mentioning above, in the second paragraph.

    ```cpp
    // This code is outside TensorFlow
    size_t ExternalFileSize(const string& path) {
      RandomAccessFile file;
      Env::Default()->NewRandomAccessFile(path, &file);
      return file->GetSize();
   }

   // This code is in plugin (both A and B)
   size_t GetFileSize(const TF_RandomAccessFile *file) {
      const char *filename = file->filename;

      // SizeAtStart, SizeAtEnd, ReadFirstBlock, ReadLastBlock, ReadSizeFromMetadata, ReadFromStartToEndAndGetSize
      // are all functions implemented in each plugin, code duplication
      if (SizeAtStart(filename)) {
        char* metadata = ReadFirstBlock(file);
        return ReadSizeFromMetadata(metadata);
      } else if (SizeAtEnd(filename)) {
        char* metadata = ReadLastBlock(file);
        return ReadSizeFromMetadata(metadata);
      } else {
        // This calls the plugin function, not the external one!
        return ReadFromStartToEndAndGetSize(file);
      }
    }
    ```

3. We allow `Read` to receive a negative argument for the offset. In this case, `ExternalFileSize(path)` first looks at the path extension. If it is `none` it calls `Read` and increments a size variable until end of file, returning this size. If it is `end`, it calls `Read` with the proper negative offset, reads until EOF and returns the size from the metadata. If it is `start`, it reads from start, only the metadata and returns the size field. Since `Read` now is on core tensorflow, each plugin only needs to implement it with support for negative offsets. If negative offsets are not supported (i.e., streaming filesystems) then `Read` with negative offset fails and code falls back to the same case as `none` extension above.

    ```cpp
    // This code is outside TensorFlow
    size_t ExternalFileSize(const string& path) {
      RandomAccessFile file;
      Env::Default()->NewRandomAccessFile(path, &file);

      // SizeAtStart, SizeAtEnd, ReadSizeFromMetadata, ReadFromStartToEndAndGetSize
      // implemented only once in the external code
      if (SizeAtStart(filename)) {
        file->Read(0, amount_to_read, output, metadata);
        return ReadSizeFromMetadata(metadata);
      } else if (SizeAtEnd(filename)) {
        if (file->Read(-amount_to_read, amount_to_read, output, metadata)) {
          return ReadSizeFromMetadata(metadata);
        } else {
          // filesystem doesn't support negative offsets (filesystem B)
          // so, fallback to default
          return ReadFromStartToEndAndGetSize(file);
        }
      } else {
        return ReadFromStartToEndAndGetSize(file);
      }
      return file->GetSize();
   }

   // This code is in plugin A
   size_t Read(const TF_RandomAccessFile *file, uint64 offset, ...) {
      if (offset < 0) { 
        // position from end and read ...
      } else {
        // read directly
        ...
      }
   }

   // This code is in plugin B
   size_t Read(const TF_RandomAccessFile *file, uint64 offset, ...) {
      if (offset < 0) { 
        // error as seek is not supported
      } else {
        // read directly
        ...
      }
    }
    ```

4. We change `Read` to only receive the number of bytes to read and add a `Seek` method to specify where the read should happen from. `Seek` will have an argument for the relative position (negative offset, positive offset) and another enum argument to specify if the position change should be from current position, start of file or end of file. 

    ```cpp
    // This code is outside TensorFlow
    size_t ExternalFileSize(const string& path) {
      RandomAccessFile file;
      Env::Default()->NewRandomAccessFile(path, &file);

      // SizeAtStart, SizeAtEnd, ReadSizeFromMetadata, ReadFromStartToEndAndGetSize
      // implemented only once in the external code
      if (SizeAtStart(filename)) {
        file->Read(amount_to_read, output, metadata);
        return ReadSizeFromMetadata(metadata);
      } else if (SizeAtEnd(filename) && file->Seek(-amount_to_read, FILE_END)) {
        file->Read(amount_to_read, output, metadata);
        return ReadSizeFromMetadata(metadata);
      } else {
        return ReadFromStartToEndAndGetSize(file);
      }
      return file->GetSize();
   }
    ```
   This is similar to 3. above, but has a clearer API.

I would go with 4 or 3 instead of 2 as you propose.

Reply:
@mihaimaruseac Thanks. I think 3 or 4 makes sense 👍 . For 3, maybe the offset will be `int64` instead of `uint64` (to allow negative numbers)?

Reply:
It has to be `int64`, indeed

Reply:
Option (**4**) would prevent concurrent reads from the RandomAccessFile (and require additional locking ...) - while this functionality is currently underused I expect this to change.

I would really like to see a GetSize() (**2**) method for RandomAccessFile -  currently GetFileSize() is used as a substituted either before or after NewRandomAccessFile().
GetSize() could ensure on most file systems that both operations actually refer to the same file object in the presence of rename/delete file operations.

Reply:
So that leaves option 3 to read the size after decompression.

Indeed, `GetFileSize()` returns the size of the file on disk. Probably we can also implement `GetSize()` which does the same thing, returning the size of the file at the moment it was read (and filesystems can decide to not implement it in which case we'll default to calling `GetFileSize()` from the filesystem object itself)